R/data.table.R

```diff
@@ -163,27 +163,36 @@ replace_dot_alias = function(e) {
   byjoin = !is.null(by) && is.symbol(bysub) && bysub==".EACHI"
   naturaljoin = FALSE
   names_x = names(x)
-  if (missing(i) && !missing(on)) {
-    tt = eval.parent(.massagei(substitute(on)))
-    if (!is.list(tt) || !length(names(tt))) {
-      warning("When on= is provided but not i=, on= must be a named list or data.table|frame, and a natural join (i.e. join on common names) is invoked. Ignoring on= which is '",class(tt)[1L],"'.")
-      on = NULL
-    } else {
-      i = tt
-      naturaljoin = TRUE
-    }
-  }
-  if (missing(i) && missing(j)) {
-    tt_isub = substitute(i)
-    tt_jsub = substitute(j)
-    if (!is.null(names(sys.call())) &&  # not relying on nargs() as it considers DT[,] to have 3 arguments, #3163
-        tryCatch(!is.symbol(tt_isub), error=function(e)TRUE) &&   # a symbol that inherits missingness from caller isn't missing for our purpose; test 1974
-        tryCatch(!is.symbol(tt_jsub), error=function(e)TRUE)) {
-      warning("i and j are both missing so ignoring the other arguments. This warning will be upgraded to error in future.")
+  if (missing(i)) {
+    if (!missing(on)) {
+      tt = eval.parent(.massagei(substitute(on)))
+      if (!is.list(tt) || !length(names(tt))) {
+        warning(domain=NA, gettextf(
+          "When on= is provided but not i=, on= must be a named list or data.table|frame, and a natural join (i.e. join on common names) is invoked. Ignoring on= which is '%s'.",
+          class(tt)[1L], domain="R-data.table"
+        ))
+        on = NULL
+      } else {
+        i = isub = tt
+        naturaljoin = TRUE
+      }
     }
-    return(x)
+    
+    if (missing(i) && missing(j)) { ## Checking again because we may have changed it above
+      tt_isub = substitute(i)
+      tt_jsub = substitute(j)
+      if (!is.null(names(sys.call())) &&  # not relying on nargs() as it considers DT[,] to have 3 arguments, #3163
+          tryCatch(!is.symbol(tt_isub), error=function(e)TRUE) &&   # a symbol that inherits missingness from caller isn't missing for our purpose; test 1974
```

Comment:
what errors are being caught here?

Reply:
Here's the test. In this case, ```cols``` is missing I believe.

```r
# no error when j is supplied but inherits missingness from caller
DT = data.table(a=1:3, b=4:6)
f = function(cols) DT[,cols]
test(1974.1, f(), output="a.*b.*3:.*6")
```